### PR TITLE
feat(rum-angular): expose centralized error handler for angular

### LINKED ISSUE #545: expose `APMErrorHandler` from Angular integration package
+ By default the Default Angular Error Handler captures the app errors and logs them in to the console without rethrowing them as browser error event, so we should encourage people to write a custom error handler on how to capture errors in angular apps and report it to APM. 

```ts
import { ErrorHandler } from '@angular/core'
import { apm } from '@elastic/apm-rum'

export class APMErrorHandler implements ErrorHandler {
  handleError(error) {
    apm.captureError(error)
  }
}

// Usage in App module 
@NgModule({
  providers: [{provide: ErrorHandler, useClass: APMErrorHandler}]
})
class MyModule {}
```

### PR DESCRIPTION
**Pull request by vigneshshanmugam**
+ fix elastic/apm-agent-rum-js#545 
+ Provides a centralised error handling for angular apps which captures and reports the error, plus also logs them to the console for debugging purposes. 
```ts
import { ErrorHandler } from '@angular/core'
import { ApmErrorHandler } from '@elastic/apm-rum-angular'
@NgModule({
  providers: [
    {
      provide: ErrorHandler,
      useClass: ApmErrorHandler
    }
  ]
})
class AppModule {}

```

### CODE REVIEW

File: docs/angular-integration.asciidoc [hunk 1]
```diff
@@ -62,3 +62,29 @@ export class AppModule {
 
 Once the service is initialized, both page load and SPA navigation events will be captured
 as transactions with the `path` of the route as its name and `page-load` or `route-change` as type.
+
+[float]
+==== Capturing errors in Angular applications
+
+By default, when an error is thrown inside the Angular application, the default error handler printss
```
Comment (bmorelli25) [inline docs/angular-integration.asciidoc]:
```suggestion
By default, when an error is thrown inside the Angular application, the default error handler prints
```

Comment (bmorelli25) [reply docs/angular-integration.asciidoc]: 🐍

**Comment by vigneshshanmugam** [comment]
@hamid I have fixed the tests, Please have a look now

File: packages/rum-angular/test/specs/apm-service.spec.ts [hunk 2]
```diff
@@ -24,7 +24,6 @@
  */
 
 import '../polyfills'
-import 'zone.js/dist/zone-testing'
```
[the remark this answers is not in the document]
Comment (vigneshshanmugam) [reply packages/rum-angular/test/specs/apm-service.spec.ts]: It was added by mistake, need only when we test Zone internals and async stuff. We don't use anything at the moment.

File: packages/rum-angular/src/error-handler.ts [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ *
+ */
+
+import { ErrorHandler } from '@angular/core'
+
+export class ApmErrorHandler implements ErrorHandler {
```
[the remark this answers is not in the document]
Comment (vigneshshanmugam) [reply packages/rum-angular/src/error-handler.ts]: Had a call and we have decided to extend the ErrorHandler, capture the error in APM and also prints to the console.

File: dev-utils/karma.js [hunk 4]
```diff
@@ -49,13 +49,14 @@ const baseConfig = {
...
     'karma-sauce-launcher',
-    'karma-failed-reporter',
     'karma-jasmine',
+    'karma-failed-reporter',
     'karma-spec-reporter',
     'karma-webpack',
     'karma-sourcemap-loader'
   ],
   client: {
+    // clearContext: false, // make Jasmine Spec Runner output visible in browsers
```
Comment (vigneshshanmugam) [inline dev-utils/karma.js]: mainly for debugging, uncomment it to enable.

File: packages/rum-angular/test/specs/error-handler.spec.ts [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ *
+ */
+import '../polyfills'
```
Comment (jahtalab) [inline packages/rum-angular/test/specs/error-handler.spec.ts]: Can't this file also use `test.ts`?

File: packages/rum-angular/test/specs/error-handler.spec.ts [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+
+  it('should capture errors on app', () => {
+    expect(errorHandler).toBeInstanceOf(ApmErrorHandler)
+    spyOn(apm, 'captureError')
+    /**
+     * We extend the angular default error handler which logs
+     * the error to console, so we fake that in this test
+     * to avoids logs in stdout
+     */
+    spyOn(ErrorHandler.prototype, 'handleError').and.callFake(() => undefined)
```
Comment (jahtalab) [inline packages/rum-angular/test/specs/error-handler.spec.ts]: Probably should clean up this after the test.

Comment (vigneshshanmugam) [reply packages/rum-angular/test/specs/error-handler.spec.ts]: All spies are cleared up after each test, only stubs are kept as it is.

File: packages/rum-angular/test/specs/error-handler.spec.ts (hunk 5 above)
Comment (vigneshshanmugam) [reply packages/rum-angular/test/specs/error-handler.spec.ts]: Test.ts is the one which includes all the `.spec.ts` files.

**Comment by vigneshshanmugam** [comment]
@jahtalab Before each test is run, we have to set up the Angular testbed configuration. we cannot do that in beginning of each test as it will throw errors. So here `test.ts` acts as single place where we set up the testbed config and let test.ts require all `.spec.ts` files to be tested which happens on this line  https://github.com/elastic/apm-agent-rum-js/blob/8093bb036667ae7b08889180b0f9bd41d52e6fea/packages/rum-angular/test/specs/test.ts#L43

It is the approach followed by Angular CLI `ng test`. I tried working around having it individual files, but cant satisfy zone.js issues. So followed the Angular recommendations. 

Hope this explanation helps.

File: packages/rum-angular/karma.conf.js [hunk 7]
```diff
@@ -30,11 +30,21 @@ const {
...
+
 module.exports = function(config) {
   config.set(baseConfig)
-  config.set({
+  const preparedConfig = {
+    ...prepareConfig(config, 'rum-angular'),
+    files: [specFiles],
+    preprocessors: {
+      [specFiles]: ['webpack', 'sourcemap']
+    },
```
[the remark this answers is not in the document]
Comment (vigneshshanmugam) [reply packages/rum-angular/karma.conf.js]: Explained about why its this way https://github.com/elastic/apm-agent-rum-js/pull/559#issuecomment-573039284

**Pull request merged**